Adds SigV4 capability

harshavamsi opened the pull request:
### Description
Adds SigV4 signing capabilities to Hive, Spark, MR, Storm, and Pig.


### Issues Resolved
Closes #28

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

harshavamsi commented:
@dblock @nknize appreciate you help and input here. 

I keep running into this issue when I build and do a dry run with HIVE and SigV4. Can't seem to understand why there is a classNotFound error when the JAR is being included.

```java
Exception in thread "main" java.lang.NoClassDefFoundError: software/amazon/awssdk/auth/credentials/AwsCredentialsProvider
        at org.opensearch.hadoop.rest.commonshttp.CommonsHttpTransport.awsSigV4SignRequest(CommonsHttpTransport.java:755)
        at org.opensearch.hadoop.rest.commonshttp.CommonsHttpTransport.execute(CommonsHttpTransport.java:715)
        at org.opensearch.hadoop.rest.NetworkClient.execute(NetworkClient.java:126)
        at org.opensearch.hadoop.rest.RestClient.execute(RestClient.java:433)
        at org.opensearch.hadoop.rest.RestClient.execute(RestClient.java:429)
        at org.opensearch.hadoop.rest.RestClient.execute(RestClient.java:397)
        at org.opensearch.hadoop.rest.RestClient.mainInfo(RestClient.java:694)
        at org.opensearch.hadoop.rest.InitializationUtils.discoverClusterInfo(InitializationUtils.java:430)
        at org.opensearch.hadoop.hive.HiveUtils.init(HiveUtils.java:207)
        at org.opensearch.hadoop.hive.OpenSearchHiveInputFormat.getSplits(OpenSearchHiveInputFormat.java:122)
        at org.opensearch.hadoop.hive.OpenSearchHiveInputFormat.getSplits(OpenSearchHiveInputFormat.java:61)
        at org.apache.hadoop.hive.ql.exec.FetchOperator.generateWrappedSplits(FetchOperator.java:425)
        at org.apache.hadoop.hive.ql.exec.FetchOperator.getNextSplits(FetchOperator.java:395)
        at org.apache.hadoop.hive.ql.exec.FetchOperator.getRecordReader(FetchOperator.java:314)
        at org.apache.hadoop.hive.ql.exec.FetchOperator.getNextRow(FetchOperator.java:540)
        at org.apache.hadoop.hive.ql.exec.FetchOperator.pushRow(FetchOperator.java:509)
        at org.apache.hadoop.hive.ql.exec.FetchTask.fetch(FetchTask.java:146)
        at org.apache.hadoop.hive.ql.Driver.getResults(Driver.java:2691)
        at org.apache.hadoop.hive.ql.reexec.ReExecDriver.getResults(ReExecDriver.java:229)
        at org.apache.hadoop.hive.cli.CliDriver.processLocalCmd(CliDriver.java:259)
        at org.apache.hadoop.hive.cli.CliDriver.processCmd(CliDriver.java:188)
        at org.apache.hadoop.hive.cli.CliDriver.processLine(CliDriver.java:402)
        at org.apache.hadoop.hive.cli.CliDriver.executeDriver(CliDriver.java:821)
        at org.apache.hadoop.hive.cli.CliDriver.run(CliDriver.java:759)
        at org.apache.hadoop.hive.cli.CliDriver.main(CliDriver.java:683)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at org.apache.hadoop.util.RunJar.run(RunJar.java:323)
        at org.apache.hadoop.util.RunJar.main(RunJar.java:236)
Caused by: java.lang.ClassNotFoundException: software.amazon.awssdk.auth.credentials.AwsCredentialsProvider
        at java.net.URLClassLoader.findClass(URLClassLoader.java:387)
        at java.lang.ClassLoader.loadClass(ClassLoader.java:418)
        at java.lang.ClassLoader.loadClass(ClassLoader.java:351)
        ... 31 more
```

reta commented on build.gradle:
```diff
@@ -11,6 +11,13 @@ allprojects {
       duplicatesStrategy = DuplicatesStrategy.EXCLUDE
     }
+
+    repositories {
```
I am curious why do we need that? All artifacts has to be pulled from official repos

harshavamsi replied:
This was a failed attempt of mine to get my IDE to recognize the thirdparty shaded jars, will revert this change.

reta commented on mr/src/main/java/org/opensearch/hadoop/rest/commonshttp/CommonsHttpTransport.java:
```diff
@@ -635,13 +654,19 @@ public Response execute(Request request) throws IOException {
             entityMethod.setRequestEntity(new BytesArrayRequestEntity(ba));
             entityMethod.setContentChunked(false);
+
+            ByteArrayOutputStream outStream = new ByteArrayOutputStream();
```
```suggestion
            try (ByteArrayOutputStream outStream = new ByteArrayOutputStream()) {
                 ba.writeTo(outStream);
                 bodyBytes = outStream.toByteArray();
            }
```

reta commented:
> I keep running into this issue when I build and do a dry run with HIVE and SigV4. Can't seem to understand why there is a classNotFound error when the JAR is being included.

Those dependencies (below) are needed in classpath along with the Hadoop client. 

```
    implementation("software.amazon.awssdk:aws-core:${project.ext.awsSdkVersion}")
    implementation("software.amazon.awssdk:auth:${project.ext.awsSdkVersion}")
    implementation("software.amazon.awssdk:utils:${project.ext.awsSdkVersion}")
```

Alternatively, you could shade them (and relocate) the same way you did here https://github.com/opensearch-project/opensearch-hadoop/pull/48/files#diff-b7b046db19a707ccd2cec27ba2fdcd1a33ab3219125d95c187459c9cc26888cc:

```
dependencies {
    ...
    shaded("software.amazon.awssdk:aws-core:${project.ext.awsSdkVersion}")
    shaded("software.amazon.awssdk:auth:${project.ext.awsSdkVersion}")
    shaded("software.amazon.awssdk:utils:${project.ext.awsSdkVersion}")
}

shadowJar {
   ...
    relocate 'software.amazon', 'org.opensearch.hadoop.thirdparty.software.amazon'
}
```

reta commented on mr/src/main/java/org/opensearch/hadoop/rest/commonshttp/auth/aws/AwsSigV4Signer.java:
```diff
@@ -0,0 +1,236 @@
+import software.amazon.awssdk.auth.credentials.DefaultCredentialsProvider;
+import software.amazon.awssdk.utils.http.SdkHttpUtils;
+
+public class AwsSigV4Signer {
```
It seems like `software.amazon.awssdk.auth.signer.Aws4Signer` already supports [1] that out of the box? 

```
Aws4SignerParams signerParams = Aws4SignerParams.builder()
                .awsCredentials(credentials.resolveCredentials())
                .signingName(this.signingServiceName)
                .signingRegion(signingRegion)
                .build();
        return Aws4Signer.create().sign(req.build(), signerParams);
```

[1] https://github.com/opensearch-project/opensearch-java/pull/324/files#diff-13b760e273e29a17fa88085ec653c7e9324e23dea21ff90c718adb6b9f25f3b0R385

harshavamsi replied:
Thank you for this. I'll update my PR and try the shaded JAR route. It did not work previously but I'll give it another try.

harshavamsi commented:
@reta I updated the code to use the native signer with `Aws4Signer` but I get an error with build now that I cannot understand -- https://github.com/opensearch-project/opensearch-hadoop/actions/runs/4018598709/jobs/6904413460#step:7:211. What am I doing wrong?

reta commented:
> @reta I updated the code to use the native signer with `Aws4Signer` but I get an error with build now that I cannot understand -- https://github.com/opensearch-project/opensearch-hadoop/actions/runs/4018598709/jobs/6904413460#step:7:211. What am I doing wrong?

@harshavamsi please add `implementation("software.amazon.awssdk:sdk-core:${project.ext.awsSdkVersion}")` to your dependencies

harshavamsi commented:
> > @reta I updated the code to use the native signer with `Aws4Signer` but I get an error with build now that I cannot understand -- https://github.com/opensearch-project/opensearch-hadoop/actions/runs/4018598709/jobs/6904413460#step:7:211. What am I doing wrong?
> 
> @harshavamsi please add `implementation("software.amazon.awssdk:sdk-core:${project.ext.awsSdkVersion}")` to your dependencies

Thanks! This fixed that problem. I updated the aws sdk to use shaded jars but I still get this error 

```
Exception in thread "main" java.lang.BootstrapMethodError: java.lang.NoClassDefFoundError: org/opensearch/hadoop/thirdparty/amazon/awssdk/profiles/ProfileFile
```

reta commented:
> Thanks! This fixed that problem. I updated the aws sdk to use shaded jars but I still get this error

You need to shade profiles as well (and may be a few more modules):

```
shaded("software.amazon.awssdk:profiles:${project.ext.awsSdkVersion}")
```

harshavamsi commented:
> > Thanks! This fixed that problem. I updated the aws sdk to use shaded jars but I still get this error
> 
> You need to shade profiles as well (and may be a few more modules):
> 
> ```
> shaded("software.amazon.awssdk:profiles:${project.ext.awsSdkVersion}")
> ```

Ah okay. I added 

```   
shaded("software.amazon.awssdk:profiles:${project.ext.awsSdkVersion}")
    shaded("software.amazon.awssdk:protocols:${project.ext.awsSdkVersion}")
```

I still get 

```
Exception in thread "main" java.lang.NoClassDefFoundError: org/opensearch/hadoop/thirdparty/amazon/awssdk/protocols/jsoncore/JsonNodeParser
```

even though I added protocols.

harshavamsi commented:
> > > Thanks! This fixed that problem. I updated the aws sdk to use shaded jars but I still get this error
> > 
> > 
> > You need to shade profiles as well (and may be a few more modules):
> > ```
> > shaded("software.amazon.awssdk:profiles:${project.ext.awsSdkVersion}")
> > ```
> 
> Ah okay. I added
> 
> ```
> shaded("software.amazon.awssdk:profiles:${project.ext.awsSdkVersion}")
>     shaded("software.amazon.awssdk:protocols:${project.ext.awsSdkVersion}")
> ```
> 
> I still get
> 
> ```
> Exception in thread "main" java.lang.NoClassDefFoundError: org/opensearch/hadoop/thirdparty/amazon/awssdk/protocols/jsoncore/JsonNodeParser
> ```
> 
> even though I added protocols.

@reta i was able to fix this by including  `shaded("software.amazon.awssdk:json-utils:${project.ext.awsSdkVersion}")`

But now it's `org/opensearch/hadoop/thirdparty/amazon/awssdk/thirdparty/jackson/core/JsonFactory` that's missing. I wonder if the dependencies are colliding?

harshavamsi commented:
> > > > Thanks! This fixed that problem. I updated the aws sdk to use shaded jars but I still get this error
> > > 
> > > 
> > > You need to shade profiles as well (and may be a few more modules):
> > > ```
> > > shaded("software.amazon.awssdk:profiles:${project.ext.awsSdkVersion}")
> > > ```
> > 
> > 
> > Ah okay. I added
> > ```
> > shaded("software.amazon.awssdk:profiles:${project.ext.awsSdkVersion}")
> >     shaded("software.amazon.awssdk:protocols:${project.ext.awsSdkVersion}")
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > I still get
> > ```
> > Exception in thread "main" java.lang.NoClassDefFoundError: org/opensearch/hadoop/thirdparty/amazon/awssdk/protocols/jsoncore/JsonNodeParser
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > even though I added protocols.
> 
> @reta i was able to fix this by including `shaded("software.amazon.awssdk:json-utils:${project.ext.awsSdkVersion}")`
> 
> But now it's `org/opensearch/hadoop/thirdparty/amazon/awssdk/thirdparty/jackson/core/JsonFactory` that's missing. I wonder if the dependencies are colliding?

https://github.com/aws/aws-sdk-java-v2/blob/303a375e8edbe117f8f5a34a554e1b3830def460/third-party/third-party-jackson-core/pom.xml -- looks like jackson itself is a third-party shaded dependency.

reta commented:
> [third-party-jackson-core/pom.xml](https://github.com/aws/aws-sdk-java-v2/blob/303a375e8edbe117f8f5a34a554e1b3830def460/third-party/third-party-jackson-core/pom.xml) -- looks like jackson itself is a third-party shaded dependency.

Here, you may need it as well: https://mvnrepository.com/artifact/software.amazon.awssdk/third-party-jackson-core

harshavamsi commented:
@reta I've tested that this works as expected with the managed service. This is now complete. Would gladly accept feedback,

reta commented on mr/src/main/java/org/opensearch/hadoop/rest/RestClient.java:
```diff
@@ -520,7 +520,7 @@ public boolean indexExists(String index) {
 
     private boolean exists(String indexOrType) {
-        Request req = new SimpleRequest(HEAD, null, indexOrType);
+        Request req = new SimpleRequest(GET, null, indexOrType);
```
Hm ... the `HEAD` sounded right (we just need to check existence), why it was changed to `GET`?

penghuo replied:
+1

reta commented on qa/kerberos/src/main/scala/org/opensearch/hadoop/qa/kerberos/spark/LoadToES.scala:
```diff
@@ -53,8 +53,8 @@ class LoadToES(args: Array[String]) {
 
-    df.rdd.map(row => row.getValuesMap(row.schema.fieldNames)).saveToOpenSearch(s"${resource}_rdd")
-    df.saveToOpenSearch(s"${resource}_df")
+    // df.rdd.map(row => row.getValuesMap(row.schema.fieldNames)).saveToOpenSearch(s"${resource}_rdd")
```
Why it is commented out?

harshavamsi replied:
```
[Error] /local/home/hvamsi/clients/opensearch-hadoop/qa/kerberos/src/main/scala/org/opensearch/hadoop/qa/kerberos/spark/LoadToES.scala:56:64: value saveToOpenSearch is not a member of org.apache.spark.rdd.RDD[Map[String,Nothing]]
[Error] /local/home/hvamsi/clients/opensearch-hadoop/qa/kerberos/src/main/scala/org/opensearch/hadoop/qa/kerberos/spark/LoadToES.scala:57:8: value saveToOpenSearch is not a member of org.apache.spark.sql.DataFrame
[Error] /local/home/hvamsi/clients/opensearch-hadoop/qa/kerberos/src/main/scala/org/opensearch/hadoop/qa/kerberos/spark/ReadFromES.scala:51:24: value esJsonRDD is not a member of org.apache.spark.SparkContext
[Error] /local/home/hvamsi/clients/opensearch-hadoop/qa/kerberos/src/main/scala/org/opensearch/hadoop/qa/kerberos/spark/ReadFromES.scala:53:22: value esDF is not a member of org.apache.spark.sql.SQLContext
```

These are the build errors I get, not very sure what is going on here.

reta replied:
:+1: I will take a look

reta replied:
@harshavamsi please add these imports:

```
import org.opensearch.spark._
import org.opensearch.spark.sql._
```

and uncomment the test code

reta commented on qa/kerberos/src/main/scala/org/opensearch/hadoop/qa/kerberos/spark/ReadFromES.scala:
```diff
@@ -48,12 +48,12 @@ class ReadFromES(args: Array[String]) {
     val dsOutputDir = s"${args(0)}DS"
 
-    spark.sparkContext.esJsonRDD(s"${resource}_rdd").saveAsTextFile(rddOutputDir)
+    // spark.sparkContext.esJsonRDD(s"${resource}_rdd").saveAsTextFile(rddOutputDir)
```
Same question ...

reta replied:
@harshavamsi please add these imports:

```
import org.opensearch.spark._
import org.opensearch.spark.sql._
```

and uncomment the test code

VachaShah commented on mr/src/test/resources/log4j2.properties:
```diff
@@ -77,4 +77,9 @@ logger.Zookeeper.level=warn
+
+
+
+logger.category.org.openearch.hadoop.mr=DEBUG
```
I think this was just for debugging, can we remove it?

harshavamsi replied:
Removed debugging.

VachaShah commented on mr/src/main/java/org/opensearch/hadoop/rest/commonshttp/CommonsHttpTransport.java:
```diff
@@ -712,12 +793,95 @@ public Object run() throws Exception {
+
+        SdkHttpFullRequest signedRequest = Aws4Signer.create().sign(req.build(), signerParams);
+
+        final ImmutableMap.Builder<String, String> signerHeaders = ImmutableMap.builder();
```
May be have this method return a signed request and let the caller logic take care of adding headers to `http`?

VachaShah commented on mr/src/main/java/org/opensearch/hadoop/rest/commonshttp/CommonsHttpTransport.java:
```diff
@@ -683,7 +758,12 @@ public Response execute(Request request) throws IOException {
+        }
+
+        if (settings.getAwsSigV4Enabled()) {
+            awsSigV4SignRequest(request, http, bodyBytes);
```
Also add a test in `CommonsHttpTransportTests`?

harshavamsi replied:
Added a test for the SigV4 signer.

VachaShah commented on spark/sql-30/build.gradle:
```diff
@@ -189,6 +189,4 @@ sparkVariants {
         }
     }
-}
-sourceCompatibility = JavaVersion.VERSION_14
```
Why is this removed?

harshavamsi replied:
Reverted this change.

VachaShah reviewed (commented):
Does this change work for both service names "es" and "aoss"?

penghuo commented on mr/src/main/java/org/opensearch/hadoop/rest/commonshttp/CommonsHttpTransport.java:
```diff
@@ -712,12 +793,95 @@ public Object run() throws Exception {
+
+        Region signingRegion = Region.of(awsRegion);
+
+        final AwsCredentialsProvider credentials = DefaultCredentialsProvider.create();
```
could we provide configuration also? similar to s3a,  https://hadoop.apache.org/docs/stable/hadoop-aws/tools/hadoop-aws/index.html#Authenticating_with_S3

e.g.
```
opensearch.aws.credentials.provider: com.amazonaws.auth.InstanceProfileCredentialsProvider
```

harshavamsi replied:
My assumption was that using the [DefaultCredentialsProvider](https://sdk.amazonaws.com/java/api/latest/software/amazon/awssdk/auth/credentials/DefaultCredentialsProvider.html) covers all the possible scenarios for credentials including `InstanceProfileCredentialsProvider`

penghuo replied:
I think we should leave the decision to user.

penghuo commented on thirdparty/build.gradle:
```diff
@@ -23,10 +23,22 @@ dependencies {
     shaded("org.codehaus.jackson:jackson-mapper-asl:${project.ext.jacksonVersion}")
     shaded("org.codehaus.jackson:jackson-core-asl:${project.ext.jacksonVersion}")
+    shaded("com.google.guava:guava:${project.ext.googleGuavaVersion}")
+    shaded("software.amazon.awssdk:sdk-core:${project.ext.awsSdkVersion}")
```
Is there any guidance which package should be putted into shadow jar? I think depend on [aws-java-sdk-bundle ](https://aws.amazon.com/blogs/developer/java-sdk-bundle/) is [more generic approach](https://hadoop.apache.org/docs/stable/hadoop-aws/tools/hadoop-aws/index.html#Getting_Started).

harshavamsi replied:
I think we would need to package the dependencies in the jar either way, looks like with shadow jar we control the exact dependencies? I could be wrong, @reta any input?

penghuo replied:
aws-java-sdk-bundle is the shadowjar, it's the generic runtiime depedency for spark. opensearch-hadoop do not need to shaowjar aws-java-sdk-bundle.

reta replied:
@penghuo I think the intent here was to have `opensearch-hadoop/mr` JAR self-sufficient: the AWS dependency are not bundled with Hadoop or/and Spark, if we don't shade - we have to make them available to these runtimes somehow (@harshavamsi I think you have an exception shared on the pull request). I haven't looked into `aws-java-sdk-bundle`, will do that tomorrow.

penghuo replied:
@reta totally agree. aws-java-sdk-bundle is the shaded jar. Hadoop s3 depend on it https://hadoop.apache.org/docs/stable/hadoop-aws/tools/hadoop-aws/index.html#Getting_Started .

penghuo replied:
add sample code to explain it. https://github.com/penghuo/opensearch-hadoop/commit/098cfd9865e45f508fc5bc6d4c93b254ee46d49f#diff-4f0cebafafd8b671aaf175130520fa2953152f3637bedfaa9191441462f6d93c

```
import org.apache.spark.sql.SQLContext
import org.apache.spark.sql.SQLContext._
import org.opensearch.spark.sql._
val sql = new SQLContext(sc)
                         
val options = Map("pushdown" -> "true",     
"opensearch.nodes" -> "https://search-mytest-00001-huifxbwhdgfy7goat3mietmxne.us-west-2.es.amazonaws.com", 
"opensearch.aws.sigv4.enabled" -> "true",
"opensearch.aws.sigv4.region" -> "us-west-2",
"opensearch.nodes.resolve.hostname" -> "false",
"opensearch.net.ssl" -> "true")                         

val es = sql.esDF("accounts", options)
```

we only need to add aws-java-sdk-bundle to spark/jars, no shaded jar required.

reta replied:
@penghuo thanks! So what we could do is 2 things:
 - document that  `aws-java-sdk-bundle` has to be added to runtime classpath (fe spark/jars)
 - refactor the code so the AwsSig is becoming optional, in this case if it is not used, the `aws-java-sdk-bundle` won't be required

reta replied:
You have to copy it manually

harshavamsi replied:
@reta I add `implementation 'com.amazonaws:aws-java-sdk-bundle:1.12.397'` [here](https://github.com/opensearch-project/opensearch-hadoop/pull/48/files#diff-e1955902ceda6d0d39c7891161eb56eadfde6d1a78897908ab7043af09105258R109) but it does not seem to help with the `NoClassDef` error

reta replied:
I mean you have to manually copy the com.amazonaws:aws-java-sdk-bundle:1.12.397 jar to spark/jars, as per @penghuo comment https://github.com/opensearch-project/opensearch-hadoop/pull/48#discussion_r1094103160

penghuo replied:
@harshavamsi summarize my idea in https://github.com/penghuo/opensearch-hadoop/commit/0feda06f0e9c36834d3cd3cf150cb24d111d3e09

it is not production code, only for explain the idea.

## Test Setup
* copy the follow jars to spark/jar folder
```
 ~/spark-3.2.1-bin-hadoop3.2  ls jars/aws-java-sdk-bundle-1.11.375.jar
jars/aws-java-sdk-bundle-1.11.375.jar
 ~/spark-3.2.1-bin-hadoop3.2  ls jars/opensearch-spark-30_2.12-3.0.0-SNAPSHOT.jar
jars/opensearch-spark-30_2.12-3.0.0-SNAPSHOT.jar
```
* configure AWS credentials 

## Test code

* Write to AOS

```
import org.apache.spark.sql.SQLContext
import org.apache.spark.sql.SQLContext._
import org.opensearch.spark.sql._

val sql = new SQLContext(sc)

val accountsRead = sql.read.json("/Users/penghuo/release/maximus/data/accounts.json")
val options = Map("pushdown" -> "true",     
"opensearch.nodes" -> "https://aos-2.3-domain", 
"opensearch.aws.sigv4.enabled" -> "true",
"opensearch.aws.sigv4.region" -> "us-west-2",
"opensearch.nodes.resolve.hostname" -> "false",
"opensearch.nodes.wan.only" -> "true",
"opensearch.net.ssl" -> "true")   
accountsRead.saveToOpenSearch("accounts-00001", options)
```

* Load from AOS
```
import org.apache.spark.sql.SQLContext
import org.apache.spark.sql.SQLContext._
import org.opensearch.spark.sql._
val sql = new SQLContext(sc)
                         
val options = Map("pushdown" -> "true",     
"opensearch.nodes" -> "https://aos-2.3-domain", 
"opensearch.aws.sigv4.enabled" -> "true",
"opensearch.aws.sigv4.region" -> "us-west-2",
"opensearch.nodes.resolve.hostname" -> "false",
"opensearch.nodes.wan.only" -> "true",
"opensearch.net.ssl" -> "true")                         

val es = sql.esDF("accounts", options)
```

penghuo commented on mr/src/main/java/org/opensearch/hadoop/rest/commonshttp/CommonsHttpTransport.java:
```diff
@@ -712,12 +793,95 @@ public Object run() throws Exception {
+        return '?' + AMPERSAND_JOINER.join(result.build());
+    }
+
+    private void awsSigV4SignRequest(Request request, HttpMethod http, byte[] bodyBytes)
```
could we reuse some work in https://github.com/opensearch-project/opensearch-java/blob/main/java-client/src/main/java/org/opensearch/client/transport/aws/AwsSdk2Transport.java.

penghuo commented on gradle.properties:
```diff
@@ -30,7 +30,7 @@ spark13Version = 1.6.2
 spark22Version = 2.2.3
 spark24Version = 2.4.4
-spark30Version = 3.2.1
+spark30Version = 3.2.3
```
Why bump spark version? [Because spark 3.2.3 release?](https://spark.apache.org/releases/spark-release-3-2-3.html). If so, I feel it could be in seperate PR.

harshavamsi replied:
Yes, looks like `3.2.2` is no longer being published.

```
> Could not resolve all files for configuration ':qa:kerberos:downloadSpark#3.2.2'.
   > Could not find spark-3.2.2-spark-3.2.2-bin-hadoop3.2.tgz (spark:spark-3.2.2:spark-3.2.2-bin-hadoop3.2).
     Searched in the following locations:
         https://apache.osuosl.org/spark/spark-3.2.2/spark-3.2.2-bin-hadoop3.2.tgz
```

Will raise a separate PR for this 👍

penghuo commented on mr/src/main/java/org/opensearch/hadoop/rest/commonshttp/CommonsHttpTransport.java:
```diff
@@ -635,13 +667,21 @@ public Response execute(Request request) throws IOException {
             entityMethod.setRequestEntity(new BytesArrayRequestEntity(ba));
             entityMethod.setContentChunked(false);
+
+            try (ByteArrayOutputStream outStream = new ByteArrayOutputStream()) {
+                ba.writeTo(outStream);
+                bodyBytes = outStream.toByteArray();
+                outStream.close();
+            }
```
1. We should avoid copy the input ByteSequence.
2. https://github.com/opensearch-project/opensearch-hadoop/blob/main/mr/src/main/java/org/opensearch/hadoop/util/BytesArrayPool.java#L54 does not implement writeTo, is there any issue?

My suggestion is add InputStream toInputStream() interface to ByteSequence class.

penghuo commented on mr/src/main/java/org/opensearch/hadoop/rest/commonshttp/CommonsHttpTransport.java:
```diff
@@ -214,7 +221,8 @@ public CommonsHttpTransport(Settings settings, SecureSettings secureSettings, St
         this.settings = settings;
         this.secureSettings = secureSettings;
-        this.clusterName = settings.getClusterInfoOrUnnamedLatest().getClusterName().getName(); // May be a bootstrap client.
+        this.clusterName = settings.getClusterInfoOrUnnamedLatest().getClusterName().getName(); // May be a bootstrap
```
is it possible to revert code format related change? a lot of code format change make it hard to tell what is the required change.
we can enforce the format in another PR.

harshavamsi commented:
@penghuo @reta @VachaShah any more suggestions?

reta commented on .github/workflows/build-mr.yml:
```diff
@@ -44,4 +44,8 @@ jobs:
\ No newline at end of file
+        run: ./gradlew opensearch-hadoop-mr:build
+
+      - name: Cat logs
```
Debug leftovers?

reta commented on gradle.properties:
```diff
@@ -54,3 +54,12 @@ antlrVersion = 3.4
+
+org.gradle.parallel=true
+org.gradle.daemon=true
+org.gradle.jvmargs=-Xmx12800M
```
12GB is, to be fair, a lot, why we need so much?

reta commented on mr/src/main/java/org/opensearch/hadoop/rest/commonshttp/auth/aws/AwsV4Signer.java:
```diff
@@ -0,0 +1,143 @@
+import java.io.ByteArrayOutputStream;
+import java.net.URLEncoder;
+
+public class AwsV4Signer {
```
Could you rename this class to something like `AWS4SignerSupport` or alike, it leads to confusion with `com.amazonaws.auth.AWS4Signer`

reta commented:
@harshavamsi LGTM, thank you!

VachaShah reviewed (approved):
LGTM! 

In a separate PR, lets also add the documentation for signing requests.
